Battle action and states fixes - Previous Albeleon 2.21 / 2.41 from battle 7

fmatthew5876 opened the pull request:
A simpler reimplementation of Albeleon 2.21.

To test these, create a 1x 1+ turn battle event which does:

```
Cure provoke, berserk, death, shock
Show Choice:
  Death: Add dead
  Provoke: Add provoke
  Confuse: Add confuse
  Shock: Add shock
  Else:
```

To test the effects of multiple states, do the show choice twice.

To test state recovery, you can use the above event. You can also inflict states, and the have a very fast actor use a recovery item.

fmatthew5876 commented:
This one is tested and ready. I have nothing more to add here.

carstene1ns commented:
jenkins: test this please (newlib check).

Ghabry commented on src/scene_battle_rpg2k.cpp:
```diff
@@ -374,6 +373,9 @@ void Scene_Battle_Rpg2k::ProcessActions() {
 bool Scene_Battle_Rpg2k::ProcessBattleAction(Game_BattleAlgorithm::AlgorithmBase* action) {
 	// Order of execution of BattleActionState:
 	// ConditionHeal > Execute > Apply > (ResultPop > ResultPush) > Death > Finished.
+	if (action == nullptr) {
```
Is it possible that the action is nullptr?
If not add a noise Output::Warning here so we get bugreports.

fmatthew5876 replied:
Thats an artifact I forgot to remove from an old implementation of this I tried.

Ghabry replied:
okay then just remove it

Ghabry commented on src/scene_battle_rpg2k.cpp:
```diff
@@ -371,6 +371,11 @@ void Scene_Battle_Rpg2k::ProcessActions() {
 bool Scene_Battle_Rpg2k::ProcessBattleAction(Game_BattleAlgorithm::AlgorithmBase* action) {
+	// Immediately quit for dead actors no move. Prevents any animations or delays.
+	if (action->GetType() == Game_BattleAlgorithm::Type::NoMove && action->GetSource()->IsDead()) {
+		return true;
```
you access the action here before the nullptr check

fmatthew5876 replied:
nullptr check not needed, it's gone.

Ghabry commented on src/game_battlealgorithm.h:
```diff
@@ -355,6 +356,12 @@ class AlgorithmBase {
 	Type GetType() const;
 
+	/**
+	 * @ret The significant restriction of the source when this
```
```@return```

fmatthew5876 commented:
Addressed review comments. I'm not in favor of adding extra nullptr checks. If you call `ProcessBattleActions()` with nullptr its a bug and a crash is the best way to get attention. Not logging (likely ignored), and then returning and continuing in some wierd broken state.
